pkgs/http_profile/lib/http_profile.dart

```diff
@@ -2,8 +2,208 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:async' show StreamController, StreamSink;
+import 'dart:developer' show addHttpClientProfilingData, Timeline;
 import 'dart:io';
 
+/// Describes an event related to an HTTP request.
+final class HttpProfileRequestEvent {
+  final int _timestamp;
+  final String _name;
+
+  /// [timestamp] should be the time at which the event occurred, as a
+  /// microsecond value on the monotonic clock used by the [Timeline].
+  HttpProfileRequestEvent({required int timestamp, required String name})
+      : _timestamp = timestamp,
+        _name = name;
+
+  Map<String, dynamic> _toJson() => <String, dynamic>{
+        'timestamp': _timestamp,
+        'event': _name,
+      };
+}
+
+/// Describes proxy authentication details associated with an HTTP request.
+final class HttpProfileProxyData {
+  final String? _host;
+  final String? _username;
+  final bool? _isDirect;
+  final int? _port;
+
+  HttpProfileProxyData(
+      {String? host, String? username, bool? isDirect, int? port})
+      : _host = host,
+        _username = username,
+        _isDirect = isDirect,
+        _port = port;
+
+  Map<String, dynamic> _toJson() => <String, dynamic>{
+        if (_host != null) 'host': _host,
+        if (_username != null) 'username': _username,
+        if (_isDirect != null) 'isDirect': _isDirect,
+        if (_port != null) 'port': _port,
+      };
+}
+
+/// Describes details about an HTTP request.
+final class HttpProfileRequestData {
+  final Map<String, dynamic> _data;
+
+  final void Function() _updated;
+
+  /// The elements of [connectionInfo] can either be [String]s or [int]s.
+  set connectionInfo(Map<String, dynamic /*String|int*/ > value) {
+    _data['connectionInfo'] = value;
+    _updated();
+  }
+
+  /// The content length of the request, in bytes.
+  set contentLength(int value) {
+    _data['contentLength'] = value;
+    _updated();
+  }
+
+  /// The cookies presented to the server (in the 'cookie' header).
+  set cookies(List<String> value) {
+    _data['cookies'] = value;
```

Comment:
Are we okay with developers being able to modify `value` after setting this property? The way this is currently written, developers can insert new cookies directly into the list and bypass the call to `_updated()`.

If we're not okay with that, we should create a copy of `value` that we insert into `_data`.

Reply:
Ah, true. Done.